### ISSUE TITLE
Reactive libraries for signals 

### ISSUE THREAD

**Issue by pmros** [joaomilho/act#14]
It seems that main/signals is something like a reactive library. So why don't using a reactive library (like Rx, Bacon, most...) instead of creting a new one? I think it would be event better if Act would be reactive library agnostic (with light adapters).
I'm just a newbie, I'm just trying to know how (and why) Act works.
Thank you again.

**Comment by joaomilho** [comment]
So I have investigated plenty of libraries for that, and thought about the problem we're trying to solve. The first issue is that most libraries are pretty large, and one of my goals is to have the smallest footprint as possible. Also, these libraries do not really allow loading only specific modules, since they rely on method chaining (there's some discussion here http://lulk.in/act/docs/react/events.html). The typical Rx code will look like this:

```
const input = document.querySelector('#input')
const valueOnEnter = Rx.DOM.keyup(input)
    .filter((event) => event.keyCode === 13)
    .map((event) => event.target.value)

valueOnEnter.subscribe((value) => doSomethingWithValue(value))
```

The same goes for xtream, mostjs... TBH I'm not sure if there is any way to actually overcome this with some webpack magic, but I suppose not. The solution I've found for this issue is simply doing it in the functional way, using `compose` (or `pipe`). This way you can load specific processes _and_ compose processes in a better way. So the value on enter can be defined as:

```
const valueOnEnter = pipe(
    filter((event) => event.keyCode === 13),
    map((event) => event.target.value)
)
```

`filter`, `map` and `pipe` are imported separately. You can easily extend `valueOnEnter` in the same way:

```
const valueOnEnterWithSomethingElse = pipe(
    valueOnEnter,
    throttle(300),
    distinct(),
    ...
)
```

---

Now if we could find a lib that allows us to do that, this would be awesome. From the implementations I've seen I feel `mostjs` and `flyd` would be the one closest to Act's philosophy. Flyd even works with the same compositional principles, but when I tried including it my builds suddenly got hundreds of bytes bigger.

---

I'm certainly open to reviewing this, but given what I've said, WDYT?

**Comment by pmros** [comment]
I understand. But, what about moving main/signals to an independent repo?

**Comment by joaomilho** [comment]
Yeap, this makes a lot of sense. It could be the default, but allow usage of others, if people don't mind bigger builds (like cyclejs). Turning into milestone item...

**Comment by joaomilho** [comment]
Now this seems to fit very well https://github.com/pull-stream/pull-stream

**Comment by rickmed** [comment]
@joaomilho there's also https://github.com/paldepind/flyd. (EDIT: just read you mentioned it)

I've only browsed through the code but it seems that act signals is a pretty awesome/complete functional "stream" library on its own. Are there any limitations vs the others? maybe more functions out of the box? but that could be implemented by a separate module act-signals-plus or something....or is it just to make it a user preference?

**Comment by joaomilho** [comment]
Hi @rickmed, thanks for taking the time. Yes, the current signal lib is a little bit more than just the barebones, bit it does lack lots of things that others have. This can be good or bad, depending on the perspective. On one side, it is simpler and easy both to learn and to change, on the other it may fall short on more advanced usage. My goals in writing it were 1) to focus on the minimal set of functionality (20%) one needs for most web apps (80%) 2) to be able to include each file as needed and compose (method chaining makes composition harder & requires a full import) and 3) to make it "transparent", that is, so simple that "anyone" can understand its implementation.

**Comment by rickmed** [comment]
I understand the reasons. Meaning, I agree with all of them. The question is, if the native library is lighter, more tailored and "transparent/simpler", why would I want to use a different one? Moreover, the api is similar (map, filter, takeuntil, etc) just that the syntax is different because (AFAIU) it uses functional composition instead of OO/method composition than most of the rest. IMHO, I would much rather decouple the view layer.

**Comment by joaomilho** [comment]
yep, mainly because it opens the doors to React components... I'll give it more thought
